diff --git a/src/psmx_iox/CMakeLists.txt b/src/psmx_iox/CMakeLists.txt
--- a/src/psmx_iox/CMakeLists.txt
+++ b/src/psmx_iox/CMakeLists.txt
@@ -41,4 +41,4 @@
-  target_link_libraries(psmx_iox2 PRIVATE iceoryx2-c::static-lib)
+  target_link_libraries(psmx_iox2 PRIVATE iceoryx2-c::shared-lib)
   if(BUILD_SHARED_LIBS)
     target_link_libraries(psmx_iox2 PRIVATE ddsc)
   endif()
diff --git a/src/psmx_iox/src/psmx_iox2_impl.c b/src/psmx_iox/src/psmx_iox2_impl.c
--- a/src/psmx_iox/src/psmx_iox2_impl.c
+++ b/src/psmx_iox/src/psmx_iox2_impl.c
@@ -894,5 +894,5 @@ static bool psmx_iox2_init_writer (psmx_iox2_endpoint_t *ep, psmx_iox2_topic_t
   dds_reliability_kind_t rel;
   (void) dds_qget_reliability (qos, &rel, NULL);
-  iox2_port_factory_publisher_builder_unable_to_deliver_strategy (&publisher_builder_handle, rel == DDS_RELIABILITY_BEST_EFFORT ? iox2_unable_to_deliver_strategy_e_DISCARD_SAMPLE : iox2_unable_to_deliver_strategy_e_BLOCK);
-  
+  iox2_port_factory_publisher_builder_backpressure_strategy (&publisher_builder_handle, rel == DDS_RELIABILITY_BEST_EFFORT ? iox2_backpressure_strategy_e_DISCARD_DATA : iox2_backpressure_strategy_e_RETRY_UNTIL_DELIVERED);
+
   int e = iox2_port_factory_publisher_builder_create (publisher_builder_handle, NULL, &ep->iox2_handle.wr);
@@ -1128,7 +1128,7 @@ static bool get_loglevel_opt (const char *configstr, iox2_log_level_e *out)
     };
     for (size_t i = 0; i < sizeof (tab) / sizeof (*tab); i++)
     {
-      if (ddsrt_strcasecmp (valstr, tab[i].name))
+      if (ddsrt_strcasecmp (valstr, tab[i].name) == 0)
       {
         *out = tab[i].level;
         ddsrt_free (valstr);
@@ -1323,7 +1323,7 @@ err_instance_name:
   iox2_config_drop (config);
   ddsrt_free (instance_name);
 err_config_from_ptr:
-  ddsrt_free (psmx);
+  ddsrt_free (psmx_iox2);
 err_trivial:
   return DDS_RETCODE_ERROR;
 }
