diff --git a/3rdparty/protobuf/src/google/protobuf/stubs/mutex.h b/3rdparty/protobuf/src/google/protobuf/stubs/mutex.h
index c4599913be..4394db4d8d 100644
--- a/3rdparty/protobuf/src/google/protobuf/stubs/mutex.h
+++ b/3rdparty/protobuf/src/google/protobuf/stubs/mutex.h
@@ -121,7 +121,7 @@ class GOOGLE_PROTOBUF_CAPABILITY("mutex") PROTOBUF_EXPORT WrappedMutex {
 #if defined(__QNX__)
   constexpr WrappedMutex() = default;
 #else
-  constexpr WrappedMutex() {}
+  constexpr WrappedMutex() = default;
 #endif
   void Lock() GOOGLE_PROTOBUF_ACQUIRE() { mu_.lock(); }
   void Unlock() GOOGLE_PROTOBUF_RELEASE() { mu_.unlock(); }
